### PR TITLE
WIP: add Alpaca API to momentum scanner

### PR DESCRIPTION
**Pull request by riven314**

### CODE REVIEW

File: liualgotrader/data/alpaca.py
Comment (amor71) [inline liualgotrader/data/alpaca.py:43]: @riven314 why is the function returning an empty list vs the snapshot? Maybe I'm missing something here.

File: liualgotrader/scanners/momentum.py
Comment (amor71) [inline liualgotrader/scanners/momentum.py:91]: @riven314 maybe write log entry for that case?

File: liualgotrader/scanners/momentum.py
Comment (amor71) [inline liualgotrader/scanners/momentum.py:98]: @riven314 I think it will be cleaner if that snapshot filtering will be implemented inside the data provider, using the specifics of the payload. I don't suppose all data providers will provide the same format of a payload . So instead of having scanner w/ run_polygon vs. run_alapca etc I'm thinking maybe we need to add some filtering criteria the get-snapshot functions, that will already return filtered data? or some other mechanism?

**Review by amor71 — CHANGES_REQUESTED** [review]
please review the notes. thank you.

File: liualgotrader/data/alpaca.py
Comment (riven314) [reply liualgotrader/data/alpaca.py:49]: it was WIP by the time
just pushed a commit that implement this function

File: liualgotrader/scanners/momentum.py
Comment (riven314) [reply liualgotrader/scanners/momentum.py:99]: this filtering rule is very specific to momentum strategy so I feel like its not the responsibility for data provider.
do you mean we should add a generic method for filtering symbols in alpaca & alpaca data provider? 
method like this could be used on other kind of filtering but it may add complexity on the implementation (e.g. what should be arguments for this method, how to pass in comparison operator ... etc.)

File: liualgotrader/data/data_base.py
Comment (riven314) [reply liualgotrader/data/data_base.py]: ok, have done the change.
"consider symbols are properties w/ getter rather than a regular function" <<< I dont understand this part. Could you clarify?

File: liualgotrader/data/alpaca.py
Comment (amor71) [inline liualgotrader/data/alpaca.py:53]: @riven314 why not use concurrency to pull chunks in parallel and combine?

File: liualgotrader/data/alpaca.py
Comment (amor71) [inline liualgotrader/data/alpaca.py:63]: @riven314 if it's done in parallel, the worker thread that pulls the chunk can apply a filter function (passed externally), this would reduce the returned size as well.

File: liualgotrader/scanners/momentum.py
Comment (sigmantium) [reply liualgotrader/scanners/momentum.py:99]: I think what @amor71  is trying to imply is that the formatting of the payload should be within the data provider, that way things like momentum scanner can just call DataLoader.get_symbols and the data provider class for their selected provider will have consistent return. That way its easier for others to implement other providers like maybe IBKR or something. I think first we would need to define a class for the symbols so that we can standardise what information should be returned from a function list get_symbols. Then filtering would be in the scanner or strategy but because all data providers return the same it would eliminate the need for if blocks separating code dependent on the provider.

File: liualgotrader/data/alpaca.py
Comment (riven314) [reply liualgotrader/data/alpaca.py]: now I used ThreadPoolExecutor to achieve concurrency. Speed wise it takes ~9 sec (v.s. ~15 sec in the past)

**Comment by amor71** [comment]
@riven314 I will merge the PR now, @riven314 I think we need to create documentation that explains how the new momentum scanner is working how to extend it?

@sigmantium thank you for the comments. Pls, take a look when you have time.

**Comment by amor71** [comment]
@riven314 Actually I can't merge it due to conflicts. Can you pls merge from the master, fix conflicts, and push? I'll merge once done.

**Comment by riven314** [comment]
@amor71 I have resolved the conflict, once the test is fine. Its ready to be merged

**Pull request merged**
